src/client/Client.cc

```diff
@@ -2607,32 +2607,20 @@ void Client::handle_client_request_forward(const MConstRef<MClientRequestForward
   ceph_assert(request);
 
   /*
-   * The type of 'num_fwd' in ceph 'MClientRequestForward'
-   * is 'int32_t', while in 'ceph_mds_request_head' the
-   * type is '__u8'. So in case the request bounces between
-   * MDSes exceeding 256 times, the client will get stuck.
+   * Avoid inifinite retrying after overflow.
    *
-   * In this case it's ususally a bug in MDS and continue
-   * bouncing the request makes no sense.
-   *
-   * In future this could be fixed in ceph code, so avoid
-   * using the hardcode here.
+   * The MDS will increase the fwd count and in client side
+   * if the num_fwd is less than the one saved in request
+   * that means the MDS is an old version and overflowed of
+   * 8 bits.
    */
-  int max_fwd = sizeof(((struct ceph_mds_request_head*)0)->num_fwd);
-  max_fwd = (1 << (max_fwd * CHAR_BIT)) - 1;
   auto num_fwd = fwd->get_num_fwd();
-  if (num_fwd <= request->num_fwd || num_fwd >= max_fwd) {
-    if (request->num_fwd >= max_fwd || num_fwd >= max_fwd) {
-      request->abort(-EMULTIHOP);
-      request->caller_cond->notify_all();
-      ldout(cct, 1) << __func__ << " tid " << tid << " seq overflow"
-                    << ", abort it" << dendl;
-    } else {
-      ldout(cct, 10) << __func__ << " tid " << tid
-                     << " old fwd seq " << fwd->get_num_fwd()
-                     << " <= req fwd " << request->num_fwd
-                     << ", ignore it" << dendl;
-    }
+  if (num_fwd <= request->num_fwd || (uint32_t)num_fwd >= (uint32_t)UINT32_MAX) {
```

Comment:
does this really need to by typecasted? `UINT32_MAX` _will be_ unsigned.

Reply:
Hmmm.. Can num_fwd overflow to 0?

Reply:
> does this really need to by typecasted? `UINT32_MAX` _will be_ unsigned.

It seems no need, I will remove it if there is not warning when compiling the code.

> Hmmm.. Can num_fwd overflow to 0?
> 
> EDIT: Then this check would be buggy in that case.

No, it won't. When the `num_fwd` reaches up to `UINT32_MAX` the client will stop forwarding the request and when a MDS receives a request the `num_fwd` will always be smaller than `UINT32_MAX`. And the MDS will increase `num_fwd` and pass it back to client.

Reply:
ack